rhombus/private/list.rkt

```diff
@@ -351,108 +651,195 @@
-                           (maybe-list-tail v '#,len)
-                           #t)))
-     (generate-binding #'form-id pred args #'tail
-                       #`(#,group-tag rest-bind #,list-static-infos
+     (generate-binding #'form-id len #t args #'tail
+                       #`(#,group-tag rest-bind #,static-infos
                           (#,group-tag rest-arg ...))
-                       (if (null? args) #'values #'cdr)
+                       (make-rest-selector len #f)
                        #f)]
     [(form-id (_ arg ... rest-arg (group _::...-bind)) . tail)
      (define args (syntax->list #'(arg ...)))
      (define len (length args))
-     (define pred #`(lambda (v)
-                      (and (list? v)
-                           (maybe-list-tail v '#,len)
-                           #t)))
-     (generate-binding #'form-id pred args #'tail #'rest-arg
-                       (if (null? args) #'values #'cdr)
+     (generate-binding #'form-id len #t args #'tail #'rest-arg
+                       (make-rest-selector len #t)
                        #t)]
     [(form-id (_ arg ...) . tail)
      (define args (syntax->list #'(arg ...)))
      (define len (length args))
-     (define pred #`(lambda (v)
-                      (and (list? v)
-                           (null? (maybe-list-tail v '#,len)))))
-     (generate-binding #'form-id pred args #'tail)]))
+     (generate-binding #'form-id len #f args #'tail)]))
+
+(define-for-syntax (generate-treelist-binding form-id len or-more? args tail [rest-arg #f] [rest-selector #f]
+                                              [rest-repetition? #t])
+  (define pred #`(lambda (v)
+                   (and (treelist? v)
+                        (#,(if or-more? #'>= #'=) (treelist-length v) #,len))))
+  ((make-composite-binding-transformer "List"
+                                       pred
+                                       (for/list ([i (in-range (length args))])
+                                         #`(lambda (tl) (treelist-ref tl #,i)))
+                                       (for/list ([arg (in-list args)])
+                                         #'())
+                                       #:index-result-info? #t
+                                       #:rest-accessor rest-selector
+                                       #:rest-repetition? rest-repetition?
+                                       #:rest-to-repetition #'treelist->list
+                                       #:static-infos treelist-static-infos)
+   #`(#,form-id (parens . #,args) . #,tail)
+   rest-arg))
+
+(define-for-syntax (generate-list-binding form-id len or-more? args tail [rest-arg #f] [rest-selector #f]
+                                          [rest-repetition? #t])
+  (define pred #`(lambda (v)
+                   (and (list? v)
+                        #,(if or-more?
+                              #`(length-at-least? v #,len)
+                              #`(= (length v) #,len)))))
```

Comment:
(Something missed during conflict resolution?)  I changed this to `maybe-list-tail` in dd02e51; see the commit message for the reason.

Reply:
HF